core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -90,30 +89,252 @@
+     * one block.
+     * When EOF is reached or the current block is the last block, it is the file length.
+     */
+    private long mNextBlockStartPos;
+
+    /**
+     * Creates a new position state.
+     *
+     * @param fileLength the file length in bytes
+     * @param blockIds the block IDs, sorted by the blocks to be read in as a stream
+     * @param blockSize the block size in bytes
+     */
+    public PositionState(long fileLength, List<Long> blockIds, long blockSize) {
+      mFileLength = fileLength;
+      mBlockIds = blockIds;
+      mBlockSize = blockSize;
+
+      mPos = 0;
+      mRemaining = mFileLength;
+      computeState();
+    }
+
+    /**
+     * Increases position by increment, updates the block index and block position if necessary.
+     *
+     * @param increment the position increment, expected to be non-negative
+     */
+    public void increase(long increment) {
+      mPos += increment;
+      mRemaining -= increment;
+
+      if (mPos < mNextBlockStartPos) {
+        return;
+      }
+      if (mPos < (mNextBlockStartPos + mBlockSize)) {
+        incrementState();
+        return;
+      }
+      computeState();
+    }
+
+    /**
+     * Sets the position to pos, updates the block index and block position if necessary.
+     *
+     * @param pos the new file position, should be in a valid range from 0 to EOF
+     */
+    public void setPos(long pos) {
+      mPos = pos;
+      mRemaining = mFileLength - mPos;
+      if (isInCurrentBlock(mPos)) {
+        // In current block.
+        return;
+      }
+      if (mPos >= mNextBlockStartPos && mPos < mNextBlockStartPos + mBlockSize) {
+        // In next block.
+        incrementState();
+        return;
+      }
+      computeState();
+    }
+
+    /**
+     * @return current file position
+     */
+    public long getPos() {
+      return mPos;
+    }
+
+    /**
+     * @return current remaining bytes
+     */
+    public long getRemaining() {
+      return mRemaining;
+    }
+
+    /**
+     * @return ID of the block where the current file position belongs to
+     */
+    public long getBlockId() {
+      return mBlockId;
+    }
+
+    /**
+     * @return the start position of the block
+     */
+    public long getBlockStartPos() {
+      return mBlockStartPos;
+    }
+
+    /**
+     * @return the start position of the next block
+     */
+    public long getNextBlockStartPos() {
+      return mNextBlockStartPos;
+    }
+
+    /**
+     * @param pos the position to check
+     * @return whether the position in the current block
+     */
+    public boolean isInCurrentBlock(long pos) {
+      return pos >= getBlockStartPos() && pos < getNextBlockStartPos();
+    }
+
+    /**
+     * @return whether EOF is reached
+     */
+    public boolean isEOF() {
+      return mPos == mFileLength;
+    }
+
+    /**
+     * Sets the states to special values when EOF is reached.
+     */
+    private void setEOFState() {
```

Comment:
why is EOF a special state?

Reply:
some variables need to be set to special values, for example, `mBlockStartPos` needs to be set to file length, `mBlockId` needs to be set to `-1`, etc.